native_client/deepspeech.cc

```diff
@@ -8,45 +8,423 @@
+
+     When fininshStream() is called, we decode the accumulated logits and return
+     the corresponding transcription.
+  */
+  StreamingState* setupStream(unsigned int prealloc_frames, unsigned int sample_rate);
+  void feedAudioContent(StreamingState* ctx, const short* buffer, unsigned int buffer_size);
+  char* finishStream(StreamingState* ctx);
+
+private:
+  /**
+   * @brief Perform decoding of the logits, using basic CTC decoder or
+   *        CTC decoder with KenLM enabled
+   *
+   * @param n_frames       Number of timesteps to deal with
+   * @param logits         Matrix of logits, of dimensions:
+   *                       [n_frames][batch_size][num_classes]
+   *
+   * @return String representing the decoded text.
+   *
+   * @note This method will free @p logits.
+   */
+  char* decode(std::vector<float>& logits);
+
+  /**
+   * @brief Do a single inference step in the acoustic model, with:
+   *          input=mfcc
+   *          input_lengths=[n_frames]
+   *
+   * @param mfcc batch input data
+   * @param n_frames number of timesteps in the data
+   *
+   * @param[out] output_logits Should be large enough to fit
+   *                           aNFrames * alphabet_size floats.
+   */
+  void infer(float* mfcc, int n_frames, std::vector<float>& output_logits);
+
+  void processAudioWindow(StreamingState* ctx, short* buf, unsigned int len);
+  void processMfccWindow(StreamingState* ctx, float* buf, unsigned int len);
+  void addZeroMfccWindow(StreamingState* ctx);
+  void processBatch(StreamingState* ctx, float* buf, unsigned int len);
 };
 
+StreamingState*
+Private::setupStream(unsigned int prealloc_frames,
+                     unsigned int /*sample_rate*/)
+{
+  Status status = session->Run({}, {}, {"initialize_state"}, nullptr);
+
+  if (!status.ok()) {
+    std::cerr << "Error running session: " << status << "\n";
+    return nullptr;
+  }
+
+  StreamingState* ctx = new StreamingState;
+  const size_t num_classes = alphabet->GetSize() + 1; // +1 for blank
+
+  ctx->accumulated_logits.reserve(prealloc_frames * BATCH_SIZE * num_classes);
+
+  memset(ctx->audio_buffer, 0, sizeof(ctx->audio_buffer));
+  ctx->audio_buffer_len = 0;
+
+  memset(ctx->mfcc_buffer, 0, sizeof(ctx->mfcc_buffer));
+  ctx->mfcc_buffer_len = MFCC_FEATURES*MFCC_CONTEXT; // initial zero entries
+
+  memset(ctx->batch_buffer, 0, sizeof(ctx->batch_buffer));
+  ctx->batch_buffer_len = 0;
+
+  ctx->skip_next_mfcc = false;
+
+  return ctx;
+}
+
+void
+Private::feedAudioContent(StreamingState* ctx,
+                          const short* buffer,
+                          unsigned int buffer_size)
+{
+  while (buffer_size > 0) {
+    while (buffer_size > 0 && ctx->audio_buffer_len < AUDIO_WIN_LEN_SAMPLES) {
+      ctx->audio_buffer[ctx->audio_buffer_len] = *buffer;
+      ++ctx->audio_buffer_len;
+      ++buffer;
+      --buffer_size;
+    }
+
+    if (ctx->audio_buffer_len == AUDIO_WIN_LEN_SAMPLES) {
```

Comment:
Is there any reason we can reach out of the while to here if this condition does not hold ? As much as I read, it should not, shouldn't it?

Maybe we could just `assert()` ?

Reply:
This can happen for example if I do:

```python
ctx = setupStream()
feedAudioContent(ctx, [0.]) # feed a single sample
```

We copy the single sample to the buffer, but it's not full yet, so we shouldn't do anything.